streamer/src/sendmmsg.rs

```diff
@@ -4,155 +4,187 @@ use std::io;
+    let hdr: mmsghdr = unsafe { mem::zeroed() };
+    hdrs.push(hdr);
+
+    let addr_storage: sockaddr_storage = unsafe { mem::zeroed() };
+    addrs.push(addr_storage);
+
+    hdrs[index].msg_hdr.msg_iov = &mut iovs[index];
+    hdrs[index].msg_hdr.msg_iovlen = 1;
 
     match InetAddr::from_std(dest) {
         InetAddr::V4(addr) => {
-            let index = addr_in.len();
-            addr_in.push(addr);
-            hdr.msg_hdr.msg_name = &mut addr_in[index] as *mut _ as *mut _;
-            hdr.msg_hdr.msg_namelen = addr_in_len;
+            unsafe {
+                core::ptr::write(&mut addrs[index] as *mut _ as *mut _, addr);
+            }
+            hdrs[index].msg_hdr.msg_name = &mut addrs[index] as *mut _ as *mut _;
+            hdrs[index].msg_hdr.msg_namelen = addr_in_len;
         }
         InetAddr::V6(addr) => {
-            let index = addr_in6.len();
-            addr_in6.push(addr);
-            hdr.msg_hdr.msg_name = &mut addr_in6[index] as *mut _ as *mut _;
-            hdr.msg_hdr.msg_namelen = addr_in6_len;
+            unsafe {
+                core::ptr::write(&mut addrs[index] as *mut _ as *mut _, addr);
+            }
+            hdrs[index].msg_hdr.msg_name = &mut addrs[index] as *mut _ as *mut _;
+            hdrs[index].msg_hdr.msg_namelen = addr_in6_len;
         }
     };
-    hdr
 }
 
 #[cfg(target_os = "linux")]
-pub fn send_mmsg(sock: &UdpSocket, packets: &[(&Vec<u8>, &SocketAddr)]) -> io::Result<usize> {
-    use libc::{sendmmsg, socklen_t};
-    use std::mem;
+fn sendmmsg_retry(sock: &UdpSocket, hdrs: &mut Vec<mmsghdr>) -> io::Result<()> {
+    use libc::sendmmsg;
     use std::os::unix::io::AsRawFd;
 
-    // The vectors are allocated with capacity, as later code inserts elements
-    // at specific indices, and uses the address of the vector index in hdrs
-    let mut iovs: Vec<iovec> = Vec::with_capacity(packets.len());
-    let mut addr_in: Vec<sockaddr_in> = Vec::with_capacity(packets.len());
-    let mut addr_in6: Vec<sockaddr_in6> = Vec::with_capacity(packets.len());
-
-    let addr_in_len = mem::size_of_val(&addr_in) as socklen_t;
-    let addr_in6_len = mem::size_of_val(&addr_in6) as socklen_t;
     let sock_fd = sock.as_raw_fd();
+    let mut pktidx = 0;
+    let mut total_sent = 0;
+    let mut err = None;
+
+    while pktidx < hdrs.len() {
+        let npkts =
+            match unsafe { sendmmsg(sock_fd, &mut hdrs[0], (hdrs.len() - pktidx) as u32, 0) } {
+                -1 => {
+                    err = Some(io::Error::last_os_error());
+                    0
+                }
+                n => n as usize,
+            };
+        if npkts < hdrs.len() - pktidx {
+            debug!(
+                "sendmmsg_retry failed to send some packets {}/{} error: {:?}",
+                npkts,
+                hdrs.len() - pktidx,
+                err,
+            );
+            // skip the packet we failed to send
+            pktidx += 1;
+        }
+        pktidx += npkts;
```

Comment:
The existing code doesn't really have a coherent retry mechanism. If there was a packet with an invalid address it would fail in the sendmmsg() path, then retry and fail again in the sendmmsg() path, then fall back to sending packets individually (where it would again fail to send). The existing code is written such that some packets might get a second retry in the sendmmsg path, but once we fall back to sending individual packets all packets are only tried a single time. Given that this is UDP we only see an error if we fail to put the packet on the wire. This change ensures that we skip the failing packet and make progress. Each packet gets a single chance to be sent.

Reply:
The change now retries the failing packet one additional time. This is needed to get the failure error code.